contracts/utils/cryptography/P256.sol

```diff
@@ -0,0 +1,324 @@
+
+    /**
+     * @dev Same as {verify}, but it will revert if the required precompile is not available.
+     */
+    function verify7212(uint256 h, uint256 r, uint256 s, uint256 qx, uint256 qy) internal view returns (bool) {
+        (bool valid, bool supported) = tryVerify7212(h, r, s, qx, qy);
+        if (supported) {
+            return valid;
+        } else {
+            revert Errors.MissingPrecompile(address(0x100));
+        }
+    }
+
+    /**
+     * @dev Same as {verify}, but it will return false if the required precompile is not available.
+     */
+    function tryVerify7212(
+        uint256 h,
+        uint256 r,
+        uint256 s,
+        uint256 qx,
+        uint256 qy
+    ) internal view returns (bool valid, bool supported) {
+        (bool success, bytes memory returndata) = address(0x100).staticcall(abi.encode(h, r, s, qx, qy));
+        return (success && returndata.length == 0x20) ? (abi.decode(returndata, (bool)), true) : (false, false);
+    }
+
+    /**
+     * @dev Same as {verify}, but only the Solidity implementation is used.
+     */
+    function verifySolidity(uint256 h, uint256 r, uint256 s, uint256 qx, uint256 qy) internal view returns (bool) {
+        if (r == 0 || r >= N || s == 0 || s >= N || !isOnCurve(qx, qy)) return false;
+
+        JPoint[16] memory points = _preComputeJacobianPoints(qx, qy);
+        uint256 w = Math.invModPrime(s, N);
+        uint256 u1 = mulmod(h, w, N);
+        uint256 u2 = mulmod(r, w, N);
+        (uint256 x, ) = _jMultShamir(points, u1, u2);
+        return (x == r);
+    }
+
+    /**
+     * @dev Public key recovery
+     *
+     * @param h - hashed message
+     * @param v - signature recovery param
+     * @param r - signature half R
+     * @param s - signature half S
+     *
+     * WARNING: Signatures are malleable, and this function does not check for malleability. Consider rejecting
+     * the upper half order of the curve (i.e. s > N/2)
+     */
+    function recovery(uint256 h, uint8 v, uint256 r, uint256 s) internal view returns (uint256, uint256) {
+        if (r == 0 || r >= N || s == 0 || s >= N || v > 1) return (0, 0);
+
+        uint256 rx = r;
+        uint256 ry2 = addmod(mulmod(addmod(mulmod(rx, rx, P), A, P), rx, P), B, P); // weierstrass equation y² = x³ + a.x + b
+        uint256 ry = Math.modExp(ry2, P1DIV4, P); // This formula for sqrt work because P ≡ 3 (mod 4)
+        if (mulmod(ry, ry, P) != ry2) return (0, 0); // Sanity check
+        if (ry % 2 != v % 2) ry = P - ry;
+
+        JPoint[16] memory points = _preComputeJacobianPoints(rx, ry);
+        uint256 w = Math.invModPrime(r, N);
+        uint256 u1 = mulmod(N - (h % N), w, N);
+        uint256 u2 = mulmod(s, w, N);
+        (uint256 x, uint256 y) = _jMultShamir(points, u1, u2);
+        return (x, y);
+    }
+
+    /**
+     * @dev derivate public key
+     * @param privateKey - private key
+     */
+    function getPublicKey(uint256 privateKey) internal view returns (uint256, uint256) {
```

Comment:
Would you mind if I move this function to the tests?
I see they’re relevant to derive the public key in Solidity but I’m expecting Foundry to eventually release it in Vm where the underlying implementation is in rust

Reply:
Ok